lib/Doctrine/Common/Collections/CriteriaFactory.php

```
@@ -0,0 +1,98 @@
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license. For more information, see
+ * <http://www.doctrine-project.org>.
+ */
+
+namespace Doctrine\Common\Collections;
+
+use Doctrine\Common\Collections\Expr\Comparison;
+use Doctrine\Common\Collections\Expr\CompositeExpression;
+use Doctrine\Common\Collections\Exception;
+
+/**
+ * @author Nikola Posa <posa.nikola@gmail.com>
+ */
+class CriteriaFactory
+{
+    public function create(array $criteria)
+    {
+        $expression = $orderings = $firstResult = $maxResults = $page = null;
+
+        if (isset($criteria['expression'])) {
+            $expression = $this->buildExpression($criteria['expression']);
+        }
+
+        if (isset($criteria['orderings'])) {
+            if (! is_array($criteria['orderings'])) {
+                throw Exception\InvalidCriteriaArrayException::fromInvalidOrderings($criteria['orderings']);
+            }
+
+            $orderings = $criteria['orderings'];
+        }
+
+        if (isset($criteria['first_result'])) {
+            $firstResult = $criteria['first_result'];
+        }
+
+        if (isset($criteria['max_results'])) {
+            $maxResults = $criteria['max_results'];
+        }
+
+        return new Criteria($expression, $orderings, $firstResult, $maxResults);
+    }
+
+    protected function buildExpression($expression)
+    {
+        if (! is_array($expression)) {
+            throw Exception\InvalidCriteriaArrayException::fromInvalidExpression($expression);
+        }
+
+        if (
+            isset($expression['fld'])
+            && isset($expression['op'])
+            && isset($expression['val'])
+        ) {
+            return new Comparison($expression['fld'], $expression['op'], $expression['val']);
+        }
+
+        $expressionsArray = $expression;
+        $compositeExpressionType = CompositeExpression::TYPE_AND;
+
+        if (isset($expression['$or'])) {
```

answer today: they are silently ignored